File: ouroboros-runner/internal/conn/protocol.go
package conn

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// The ceilings this protocol line fixes. They are the numbers published as
// `$defs/limits` in schemas/runner-protocol/v1.json, and limits_test.go asserts that
// they still are — a constant copied out of a document is a constant that drifts from
// it, so the copy is checked rather than trusted.
//
// These are not the session's operating values. Those are chosen per session by the
// gateway and arrive in ack.limits, each at or below the matching ceiling here, which
// is what lets the fleet be re-tuned without a release nobody can force onto customer
// hardware.
const (
	// Version is the protocol line this build writes and the only one it reads.
	Version = 1

	// MinVersion is the oldest line this build still speaks. There is no protocol 0.
	MinVersion = 1

	// EnvelopeMaxBytes is the largest single frame. A frame over it is not a frame.
	EnvelopeMaxBytes = 65536

	// LogChunkMaxBytes is the largest DECODED log chunk.
	LogChunkMaxBytes = 32768

	// LogChunkMaxBase64Chars is what LogChunkMaxBytes becomes on the wire, because
	// base64 inflates by four characters per three bytes: ceil(32768/3)*4.
	LogChunkMaxBase64Chars = 43692
)

// Type is a message type — the closed set in the envelope's `type` field.
type Type string

// The fifteen message types of protocol 1, grouped as
// docs/RUNNER_PROTOCOL.md groups them.
const (
	TypeHello       Type = "hello"      // session — the agent's first frame
	TypeAck         Type = "ack"        // session — the gateway's answer
	TypeRefuse      Type = "refuse"     // session — the gateway's refusal, then a close
	TypeHeartbeat   Type = "heartbeat"  // liveness and telemetry
	TypeJobOffer    Type = "job.offer"  // dispatch
	TypeJobAccept   Type = "job.accept" // dispatch
	TypeJobDecline  Type = "job.decline"
	TypeJobStart    Type = "job.start"    // execution
	TypeJobProgress Type = "job.progress" // execution, advisory
	TypeJobFinish   Type = "job.finish"   // execution, TERMINAL and idempotent
	TypeLogChunk    Type = "log.chunk"    // output
	TypeReceipt     Type = "receipt"      // delivery — what lets a re-send stop
	TypeDrain       Type = "drain"        // control
	TypeUndrain     Type = "undrain"      // control
	TypeBye         Type = "bye"          // control
)

// MessageTypes is every type of protocol 1, in the order v1.json enumerates them.
//
// It is exported because it is the list the repository's own checks walk: every type
// here has a section in the protocol document, a worked example in the fixture set and
// a case in expected.json, and scripts/verify-runner-protocol.sh fails when one of
// those is missing.
var MessageTypes = []Type{
	TypeHello, TypeAck, TypeRefuse, TypeHeartbeat,
	TypeJobOffer, TypeJobAccept, TypeJobDecline,
	TypeJobStart, TypeJobProgress, TypeJobFinish,
	TypeLogChunk, TypeReceipt,
	TypeDrain, TypeUndrain, TypeBye,
}

// Terminal reports whether a frame of this type is a terminal frame: one whose delivery
// must be acknowledged, whose envelope id is its idempotency id, and which a receiver
// deduplicates. job.finish is the only one in this line, and Receipt.OfType's enum is
// what stops a second being added without the contract changing.
func (t Type) Terminal() bool { return t == TypeJobFinish }

// The diagnostic codes. An implementation reports a code and the RFC 6901 path it is
// anchored at, and renders its own prose for the human — a contract over English
// sentences is one nobody can translate.
//
// Two mappings are worth stating, because they are the ones a second implementation
// would otherwise guess at:
//
//   - CodePayloadFieldType covers a value of the wrong JSON type AND a value of the
//     right type in the wrong published shape — a job id that is not a job id, a
//     timestamp that is not RFC 3339 with milliseconds.
//   - CodePayloadFieldRange covers every bound: a number outside its minimum or
//     maximum, a string outside its length, an array outside its item count.
const (
	CodeEnvelopeMalformed          = "envelope.malformed"
	CodeEnvelopeFieldMissing       = "envelope.field.missing"
	CodeEnvelopeFieldType          = "envelope.field.type"
	CodeEnvelopeFieldUnknown       = "envelope.field.unknown"
	CodeEnvelopeVersionUnsupported = "envelope.version.unsupported"
	CodeEnvelopeTypeUnknown        = "envelope.type.unknown"
	CodeEnvelopeIDMalformed        = "envelope.id.malformed"

	CodePayloadFieldMissing  = "payload.field.missing"
	CodePayloadFieldType     = "payload.field.type"
	CodePayloadFieldUnknown  = "payload.field.unknown"
	CodePayloadFieldEnum     = "payload.field.enum"
	CodePayloadFieldRange    = "payload.field.range"
	CodePayloadFieldConflict = "payload.field.conflict"
)

// Diagnostic is one reason a frame was rejected: a stable code, and the RFC 6901 path
// it is anchored at. The empty path is the frame itself.
type Diagnostic struct {
	Code string `json:"code"`
	Path string `json:"path"`
}

// Diagnostics is a rejection, in the contract's order: by path — with numeric path
// segments compared as numbers rather than as text, so /0 sorts before /10 — and then
// by code. Both implementations sort identically, which is what makes expected.json a
// comparison rather than a set membership test.
type Diagnostics []Diagnostic

// Sort puts the diagnostics into the contract's order, in place.
func (d Diagnostics) Sort() {
	sort.SliceStable(d, func(i, j int) bool {
		if d[i].Path != d[j].Path {
			return pathLess(d[i].Path, d[j].Path)
		}
		return d[i].Code < d[j].Code
	})
}

// pathLess orders two RFC 6901 pointers segment by segment, comparing two numeric
// segments as numbers. A shorter pointer that is a prefix of a longer one comes first.
func pathLess(a, b string) bool {
	as, bs := strings.Split(a, "/"), strings.Split(b, "/")
	for i := 0; i < len(as) && i < len(bs); i++ {
		if as[i] == bs[i] {
			continue
		}
		an, aerr := strconv.Atoi(as[i])
		bn, berr := strconv.Atoi(bs[i])
		if aerr == nil && berr == nil {
			return an < bn
		}
		return as[i] < bs[i]
	}
	return len(as) < len(bs)
}

// Envelope is a decoded frame: the four outer keys, and the payload as the object the
// type's own contract is applied to.
//
// The envelope is FROZEN across every version of this protocol. A gateway that speaks
// only version 1 still has to read the `v` of a version 2 hello in order to refuse it
// politely, and an agent that has just been refused still has to read the refusal. That
// is only possible if these four keys never change meaning — so a fifth key is an error
// rather than an extension point, and everything new goes in the payload, where it is
// versioned.
type Envelope struct {
	V       int
	Type    Type
	ID      string
	Payload map[string]any
}

var (
	ulidRe      = regexp.MustCompile(`^[0-9A-HJKMNP-TV-Z]{26}$`)
	sessionRe   = regexp.MustCompile(`^sess_[0-9A-HJKMNP-TV-Z]{26}$`)
	runnerRe    = regexp.MustCompile(`^rnr_[0-9A-HJKMNP-TV-Z]{26}$`)
	jobRe       = regexp.MustCompile(`^job_[0-9A-HJKMNP-TV-Z]{26}$`)
	commitRe    = regexp.MustCompile(`^[0-9a-f]{40}$`)
	timestampRe = regexp.MustCompile(`^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{3}Z$`)
)

// ValidID reports whether a string is a ULID in the published shape — what an envelope
// id, and so an outbox file's name, must be.
func ValidID(id string) bool { return ulidRe.MatchString(id) }

// envelopeKeys is the frozen set, used to report a fifth one.
var envelopeKeys = map[string]bool{"v": true, "type": true, "id": true, "payload": true}

// Decode reads one frame and returns it with every reason it is not a legal frame.
//
// An envelope error ENDS THE WALK. A frame whose envelope does not decode has no
// payload to check — reporting a payload diagnostic as well would mean guessing which
// message type the payload was meant to be, and a guess is what this contract exists to
// remove. So a rejected envelope yields envelope diagnostics only, and a nil Envelope.
//
// A frame that decodes cleanly returns its Envelope and no diagnostics. A frame whose
// envelope is legal but whose payload is not returns BOTH: the envelope, because a
// receiver may want to log what type of frame was wrong, and the diagnostics.
func Decode(raw []byte) (*Envelope, Diagnostics) {
	var diags Diagnostics

	// A frame over the ceiling is not a frame. Checked before parsing, because the
	// point of a ceiling is to bound the work a stranger can ask for.
	if len(raw) > EnvelopeMaxBytes {
		return nil, Diagnostics{{Code: CodeEnvelopeMalformed, Path: ""}}
	}

	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber() // so 1 and 1.0 stay distinguishable, which `v` depends on
	var root any
	if err := dec.Decode(&root); err != nil {
		return nil, Diagnostics{{Code: CodeEnvelopeMalformed, Path: ""}}
	}
	// One JSON object per frame, and nothing after it. A decoder stops at the end of the
	// first value it reads, so without this a frame carrying two objects would be accepted
	// as its first one — and the second would be silently discarded, which for a
	// `job.finish` means a result nobody ever sees.
	if dec.More() {
		return nil, Diagnostics{{Code: CodeEnvelopeMalformed, Path: ""}}
	}
	frame, ok := root.(map[string]any)
	if !ok {
		return nil, Diagnostics{{Code: CodeEnvelopeMalformed, Path: ""}}
	}

	for key := range frame {
		if !envelopeKeys[key] {
			diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldUnknown, Path: "/" + escape(key)})
		}
	}

	env := &Envelope{}

	switch value, present := frame["v"]; {
	case !present:
		diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldMissing, Path: "/v"})
	default:
		if n, isInt := asInteger(value); !isInt {
			diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldType, Path: "/v"})
		} else if n != Version {
			diags = append(diags, Diagnostic{Code: CodeEnvelopeVersionUnsupported, Path: "/v"})
		} else {
			env.V = n
		}
	}

	switch value, present := frame["type"]; {
	case !present:
		diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldMissing, Path: "/type"})
	default:
		name, isString := value.(string)
		switch {
		case !isString:
			diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldType, Path: "/type"})
		case specFor(Type(name)) == nil:
			diags = append(diags, Diagnostic{Code: CodeEnvelopeTypeUnknown, Path: "/type"})
		default:
			env.Type = Type(name)
		}
	}

	switch value, present := frame["id"]; {
	case !present:
		diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldMissing, Path: "/id"})
	default:
		id, isString := value.(string)
		switch {
		case !isString:
			diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldType, Path: "/id"})
		case !ulidRe.MatchString(id):
			diags = append(diags, Diagnostic{Code: CodeEnvelopeIDMalformed, Path: "/id"})
		default:
			env.ID = id
		}
	}

	switch value, present := frame["payload"]; {
	case !present:
		diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldMissing, Path: "/payload"})
	default:
		payload, isObject := value.(map[string]any)
		if !isObject {
			diags = append(diags, Diagnostic{Code: CodeEnvelopeFieldType, Path: "/payload"})
		} else {
			env.Payload = payload
		}
	}

	if len(diags) > 0 {
		diags.Sort()
		return nil, diags
	}

	spec := specFor(env.Type)
	diags = append(diags, checkObject("/payload", spec.fields, env.Payload)...)
	if spec.extra != nil {
		diags = append(diags, spec.extra(env.Payload)...)
	}
	diags.Sort()
	return env, diags
}

// Valid reports whether a frame is a legal frame of this protocol. It is the same walk
// as Decode and is here for the call sites that only want the verdict.
func Valid(raw []byte) bool {
	_, diags := Decode(raw)
	return len(diags) == 0
}

// escape encodes an object key as an RFC 6901 reference token. `~` and `/` are the two
// characters a pointer cannot carry literally.
func escape(key string) string {
	key = strings.ReplaceAll(key, "~", "~0")
	return strings.ReplaceAll(key, "/", "~1")
}

// asInteger reports the value of a JSON number with no fractional or exponent part.
// Anything else — a string, a float, a bool — is not an integer.
func asInteger(value any) (int, bool) {
	number, ok := value.(json.Number)
	if !ok {
		return 0, false
	}
	text := number.String()
	if strings.ContainsAny(text, ".eE") {
		return 0, false
	}
	n, err := strconv.Atoi(text)
	if err != nil {
		return 0, false
	}
	return n, true
}

// decodedLen is the number of bytes a base64 string decodes to, and whether it decodes
// at all. Standard encoding with padding, which is what the contract specifies.
func decodedLen(text string) (int, bool) {
	out, err := base64.StdEncoding.DecodeString(text)
	if err != nil {
		return 0, false
	}
	return len(out), true
}

File: ouroboros-rest/src/modules/farm/protocol/protocol.ts
/**
 * The runner protocol's codec — the gateway's reading of every frame an agent writes, and the
 * writing of every frame it answers with.
 *
 * AH.3 ([#251](https://github.com/NobuData/ouroboros/issues/251)), against
 * [`docs/RUNNER_PROTOCOL.md`](../../../../../docs/RUNNER_PROTOCOL.md) and
 * `schemas/runner-protocol/v1.json`. The Go agent's `internal/conn/protocol.go` is the other
 * half, and the two are held together by one file rather than by either reading the other:
 * `schemas/runner-protocol/fixtures/expected.json`, whose every case `protocol.spec.ts` decodes
 * and compares element by element — same codes, same paths, same order.
 *
 * ```
 * bytes ─▶ size ≤ 65536? ─▶ JSON object? ─▶ envelope: v · type · id · payload, nothing else
 *                                               │   (an envelope error ENDS the walk)
 *                                               ▼
 *                                         payload vs its table (protocol.schema.ts)
 *                                               ▼
 *                                    diagnostics sorted by path, then code
 * ```
 *
 * ---------------------------------------------------------------------------
 * **A diagnostic is a code and a path; never a sentence.** The contract is over the pair — § 6
 * of the protocol document — and each implementation renders its own prose. That is also why a
 * refused frame's diagnostics are safe to log: they name *where* a frame was wrong, not what
 * was in it.
 *
 * **Direction is not the codec's to check** — a codec does not know which end it runs on — so
 * {@link SENT_BY} is published here and `gateway/` applies it: a `job.offer` arriving *from* an
 * agent is a legal frame and a protocol violation.
 */

import { WireNumber, SPECS, checkObject, escape, type Diagnostic } from "./protocol.schema";
import type { Payloads } from "./protocol.messages";
import { newUlid } from "./ulid";

export { LOG_CHUNK_MAX_BASE64_CHARS, LOG_CHUNK_MAX_BYTES } from "./protocol.schema";
export type { Diagnostic, DiagnosticCode } from "./protocol.schema";

/** The newest protocol line this implementation speaks — the `v` it writes. */
export const PROTOCOL_VERSION = 1;

/** The oldest line this implementation still speaks. */
export const PROTOCOL_MIN_VERSION = 1;

/** A frame larger than this is not a frame, and is rejected without being parsed. */
export const ENVELOPE_MAX_BYTES = 65536;

/** Every message type, in `v1.json`'s order. */
export const MESSAGE_TYPES = [
  "hello",
  "ack",
  "refuse",
  "heartbeat",
  "job.offer",
  "job.accept",
  "job.decline",
  "job.start",
  "job.progress",
  "job.finish",
  "log.chunk",
  "receipt",
  "drain",
  "undrain",
  "bye",
] as const;

/** One message type. */
export type MessageType = (typeof MESSAGE_TYPES)[number];

/** Which end wrote a frame, spelt as `sessions/*.json` spells it. */
export type Sender = "agent" | "server";

/**
 * Who may send each type — fixed per type, so a frame in the wrong direction is a violation
 * and neither end has to guess whether it was meant.
 */
export const SENT_BY: Readonly<Record<MessageType, readonly Sender[]>> = {
  hello: ["agent"],
  ack: ["server"],
  refuse: ["server"],
  heartbeat: ["agent"],
  "job.offer": ["server"],
  "job.accept": ["agent"],
  "job.decline": ["agent"],
  "job.start": ["agent"],
  "job.progress": ["agent"],
  "job.finish": ["agent"],
  "log.chunk": ["agent"],
  receipt: ["server"],
  drain: ["server"],
  undrain: ["server"],
  bye: ["agent", "server"],
};

/** The terminal types — acknowledged with a `receipt`, deduplicated by envelope id. */
export const TERMINAL_TYPES: readonly MessageType[] = ["job.finish"];

/** A frame that decoded: its envelope, with the payload as plain JSON values. */
export interface Envelope<T extends MessageType = MessageType> {
  readonly v: number;
  readonly type: T;
  readonly id: string;
  readonly payload: Payloads[T];
}

/** What {@link decode} answers: the envelope when there were no diagnostics. */
export type Decoded =
  | { readonly envelope: Envelope; readonly diagnostics: readonly [] }
  | { readonly envelope: undefined; readonly diagnostics: readonly Diagnostic[] };

/** A frame about to be written. */
export interface Frame<T extends MessageType = MessageType> {
  readonly v: number;
  readonly type: T;
  readonly id: string;
  readonly payload: Payloads[T];
}

/** The four envelope keys, and the only four. */
const ENVELOPE_KEYS = new Set(["v", "type", "id", "payload"]);

/** A bare ULID — `$defs/ulid`. */
const ULID = /^[0-9A-HJKMNP-TV-Z]{26}$/;

/**
 * Decode and judge one frame.
 *
 * @param raw - The frame as it arrived: the bytes of one text message.
 * @returns The envelope and no diagnostics, or no envelope and at least one.
 */
export function decode(raw: Buffer | string): Decoded {
  const bytes = typeof raw === "string" ? Buffer.byteLength(raw) : raw.length;
  if (bytes > ENVELOPE_MAX_BYTES) return refused([{ code: "envelope.malformed", path: "" }]);

  const root = parse(typeof raw === "string" ? raw : raw.toString("utf8"));
  if (root === undefined || !isObject(root)) {
    return refused([{ code: "envelope.malformed", path: "" }]);
  }

  const diagnostics: Diagnostic[] = [];

  for (const key of Object.keys(root)) {
    if (!ENVELOPE_KEYS.has(key)) {
      diagnostics.push({ code: "envelope.field.unknown", path: `/${escape(key)}` });
    }
  }

  const v = root.v;
  if (!Object.hasOwn(root, "v")) diagnostics.push({ code: "envelope.field.missing", path: "/v" });
  else if (!(v instanceof WireNumber) || !v.integer) {
    diagnostics.push({ code: "envelope.field.type", path: "/v" });
  } else if (v.value !== PROTOCOL_VERSION) {
    diagnostics.push({ code: "envelope.version.unsupported", path: "/v" });
  }

  const type = root.type;
  if (!Object.hasOwn(root, "type")) {
    diagnostics.push({ code: "envelope.field.missing", path: "/type" });
  } else if (typeof type !== "string") {
    diagnostics.push({ code: "envelope.field.type", path: "/type" });
  } else if (!isMessageType(type)) {
    diagnostics.push({ code: "envelope.type.unknown", path: "/type" });
  }

  const id = root.id;
  if (!Object.hasOwn(root, "id")) diagnostics.push({ code: "envelope.field.missing", path: "/id" });
  else if (typeof id !== "string") diagnostics.push({ code: "envelope.field.type", path: "/id" });
  else if (!ULID.test(id)) diagnostics.push({ code: "envelope.id.malformed", path: "/id" });

  const payload = root.payload;
  if (!Object.hasOwn(root, "payload")) {
    diagnostics.push({ code: "envelope.field.missing", path: "/payload" });
  } else if (!isObject(payload)) {
    diagnostics.push({ code: "envelope.field.type", path: "/payload" });
  }

  // An envelope error ends the walk: a frame whose envelope does not decode has no payload to
  // check, and reporting one would mean guessing which type it was meant to be.
  if (diagnostics.length > 0) return refused(diagnostics);

  const messageType = type as MessageType;
  const body = payload as Record<string, unknown>;
  const spec = SPECS[messageType];

  const values = plain(body) as Record<string, unknown>;

  diagnostics.push(...checkObject("/payload", spec.fields, body));
  if (spec.extra) diagnostics.push(...spec.extra(values));

  if (diagnostics.length > 0) return refused(diagnostics);

  return {
    envelope: {
      v: PROTOCOL_VERSION,
      type: messageType,
      id: id as string,
      payload: values as unknown as Payloads[MessageType],
    },
    diagnostics: [],
  };
}

/**
 * Compose a frame this end is about to write.
 *
 * @param type - Which message.
 * @param payload - Its payload.
 * @param id - Its envelope id. Minted fresh unless the caller is re-sending a frame whose id is
 *   its identity.
 * @returns The frame.
 */
export function frame<T extends MessageType>(
  type: T,
  payload: Payloads[T],
  id: string = newUlid(),
): Frame<T> {
  return { v: PROTOCOL_VERSION, type, id, payload };
}

/**
 * Encode a frame for the wire.
 *
 * @param value - The frame.
 * @returns One JSON object, as the text of one WebSocket message.
 * @throws {RangeError} If it would exceed {@link ENVELOPE_MAX_BYTES} — an oversized frame is a
 *   bug in the writer, and sending it would only move the refusal to the other end.
 */
export function encode(value: Frame): string {
  const text = JSON.stringify(value);

  if (Buffer.byteLength(text) > ENVELOPE_MAX_BYTES) {
    throw new RangeError(
      `${value.type} frame is ${String(Buffer.byteLength(text))} bytes, ` +
        `over the ${String(ENVELOPE_MAX_BYTES)}-byte ceiling`,
    );
  }

  return text;
}

/**
 * Whether a string names a message type.
 *
 * @param value - The string.
 * @returns Whether it is one of {@link MESSAGE_TYPES}.
 */
export function isMessageType(value: string): value is MessageType {
  return (MESSAGE_TYPES as readonly string[]).includes(value);
}

/**
 * Sort diagnostics into the contract's order: by path — numeric segments compared as numbers,
 * so `/payload/command/2` precedes `/payload/command/10` — then by code.
 *
 * @param diagnostics - The diagnostics.
 * @returns A sorted copy.
 */
export function sortDiagnostics(diagnostics: readonly Diagnostic[]): Diagnostic[] {
  return [...diagnostics].sort((a, b) => {
    if (a.path !== b.path) return comparePaths(a.path, b.path);
    return a.code < b.code ? -1 : a.code > b.code ? 1 : 0;
  });
}

/**
 * Compare two RFC 6901 paths segment by segment.
 *
 * @param a - One path.
 * @param b - The other.
 * @returns Negative when `a` sorts first.
 */
function comparePaths(a: string, b: string): number {
  const as = a.split("/");
  const bs = b.split("/");

  for (let i = 0; i < as.length && i < bs.length; i += 1) {
    if (as[i] === bs[i]) continue;

    const an = /^[+-]?\d+$/.test(as[i]) ? Number(as[i]) : undefined;
    const bn = /^[+-]?\d+$/.test(bs[i]) ? Number(bs[i]) : undefined;
    if (an !== undefined && bn !== undefined) return an - bn;

    return as[i] < bs[i] ? -1 : 1;
  }

  return as.length - bs.length;
}

/**
 * A refusal, sorted.
 *
 * @param diagnostics - What was wrong.
 * @returns The decode result.
 */
function refused(diagnostics: readonly Diagnostic[]): Decoded {
  return { envelope: undefined, diagnostics: sortDiagnostics(diagnostics) };
}

/**
 * Parse JSON, keeping each number's source text.
 *
 * Node 24's `JSON.parse` hands a reviver the literal it parsed as `context.source`, which is
 * what lets `1` and `1.0` be told apart — see `protocol.schema.ts` on why they must be.
 *
 * @param text - The frame's text.
 * @returns The value with every number a {@link WireNumber}, or `undefined` when it is not JSON.
 */
function parse(text: string): unknown {
  try {
    return JSON.parse(text, (_key: string, value: unknown, context?: { source?: string }) => {
      if (typeof value !== "number") return value;

      const source = context?.source;
      return new WireNumber(
        value,
        source === undefined ? Number.isInteger(value) : /^-?\d+$/.test(source),
      );
    }) as unknown;
  } catch {
    return undefined;
  }
}

/**
 * Turn a parsed value back into plain JSON values.
 *
 * @param value - A value from {@link parse}.
 * @returns The same value with every {@link WireNumber} replaced by its number.
 */
function plain(value: unknown): unknown {
  if (value instanceof WireNumber) return value.value;
  if (Array.isArray(value)) return value.map(plain);
  if (isObject(value)) {
    return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, plain(item)]));
  }

  return value;
}

/**
 * Whether a parsed value is a JSON object — not an array, not null, not a number.
 *
 * @param value - The value.
 * @returns Whether it is an object.
 */
function isObject(value: unknown): value is Record<string, unknown> {
  return (
    typeof value === "object" &&
    value !== null &&
    !Array.isArray(value) &&
    !(value instanceof WireNumber)
  );
}
